man/hfst-fst2fst.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH HFST-FST2FST "1" "March 2012" "HFST" "User Commands"
.SH NAME
hfst-fst2fst \- =Convert transducers between binary formats
.SH SYNOPSIS
.B hfst-fst2fst
[\fIOPTIONS\fR...] [\fIINFILE\fR]
.SH DESCRIPTION
Convert transducers between binary formats
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fIINFILE\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIOUTFILE\fR
Write output transducer to OUTFILE
.SS "Conversion options:"
.TP
\fB\-f\fR, \fB\-\-format\fR=\fIFMT\fR
Write result in FMT format
.TP
\fB\-b\fR, \fB\-\-use\-backend\-format\fR
Write result in implementation format, without any HFST wrappers
.TP
\fB\-S\fR, \fB\-\-sfst\fR
Write output in (HFST's) SFST implementation
.TP
\fB\-F\fR, \fB\-\-foma\fR
Write output in (HFST's) foma implementation
.TP
\fB\-t\fR, \fB\-\-openfst\-tropical\fR
Write output in (HFST's) tropical weight (OpenFST) implementation
.TP
\fB\-l\fR, \fB\-\-openfst\-log\fR
Write output in (HFST's) log weight (OpenFST) implementation
.HP
\fB\-O\fR, \fB\-\-optimized\-lookup\-unweighted\fR Write output in the HFST optimized\-lookup implementation
.TP
\fB\-w\fR, \fB\-\-optimized\-lookup\-weighted\fR
Write output in optimized\-lookup (weighted) implementation
.TP
\fB\-Q\fR  \fB\-\-quick\fR
When converting to optimized\-lookup, don't try hard to compress
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
Format of result depends on format of INFILE
FMT must be name of a format usable by libhfst, i.e. one of the following:
{ foma, openfst\-tropical, openfst\-log, sfst,
.IP
optimized\-lookup\-weighted, optimized\-lookup\-unweighted }.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>
.PP
hfst\-fst2fst home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstFst2Fst>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.SH COPYRIGHT
Copyright \(co 2010 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.3.9
---------------------------

* Bugs fixed and new features added to hfst-proc

* Xfst rules tentatively supported in hfst-regexp2fst

* hfst-fst2strings does not produce duplicate paths any more 
  when using foma as backend

* hfst-lexc implementation format defaults to openfst-tropical
  instead of foma

Noteworthy changes in 3.3.8
---------------------------

* Fixes 3.3.7

Noteworthy changes in 3.3.7
---------------------------

* Compatible with OpenFst version 1.3.1

* Identity symbols supported in hfst-regexp2fst

Noteworthy changes in 3.3.6
---------------------------

* Include directives of foma and SFST header files changed so
  that standard include path is used instead of source path

* Compilation warnings and some memory leaks fixed

Noteworthy changes in 3.3.5
---------------------------

* hfst-fst2txt.cc also writes to dot/graphviz format

* Epenthesis added for Xerox rules parser

* hfst-compose-intersect gives warnings on missing symbols

Noteworthy changes in 3.3.4
---------------------------

* Substitution functions optimized and new substitution functions added that
  perform multiple substitutions at the same time.

* The tool hfst-substitute now implements the option --in-order.

* Updated to OpenFst version 1.2.10.

* Conversions between transducer formats are more efficient.

Noteworthy changes in 3.3.3
---------------------------

* HfstTransitionGraph and optimized lookup transducer code optimized

* Foma transducer name is not copied to the net but is left empty

* Small fixes to twolc

Noteworthy changes in 3.3.2
---------------------------

* Forward compatibility with OpenFst 1.2.8 is achieved.

* PKG_ macros removed from end user builds since they do not work nicely on
  all platforms.

Noteworthy changes in 3.3.1
---------------------------

* Handling of flag diacritics improved.

* More examples in fsmbook-examples included in tests.

* Extra files created during tests as well as generated autotools files 
  are now ignored by svn.

Noteworthy changes in 3.3.0
---------------------------

* HFST updated to work with the newest versions of backend libraries

* a new replace down function by Karttunen added

* HfstTransitionGraph and conversions between formats made less error-prone

* protected/public changes in API

* SFST and foma libraries are disabled from default build

Noteworthy changes in 3.2.0
---------------------------

* Numerous improvements to hfst-proc, including better uppercasing via glib,
  better compound handling, new analysis mode for raw analysis without case
  changes or special symbols etc.

* old slow lexc (not built by default)

* numerous scripting fixes for mac os x compatibility with foma

* constness changes in API

Noteworthy changes in 3.1.1
---------------------------

* Numerous bugfixes, in particular regarding a broken assumption in
  compose-intersect

Noteworthy changes in 3.1.0
---------------------------

* api for lookup and lookdown changed, previously memory didn't get initialized
  anywhere, causing a serious bug

Noteworthy changes in 3.0.3
---------------------------

* numerous internal improvements, eg. distcheck is now functional

* many bugfixes, especially wrt. INDENTITY_PAIR & OTHER SYMBOL, composition,
  final weights in ol-converted transducers

Noteworthy changes in 3.0.2
---------------------------

* some fixes to OSX compatibility

* no more license conflict problems wrt foma

* better memory use in compose-intersect

* new implementation of internal transducer

Noteworthy changes in 3.0.1
---------------------------

* this is a bugfix release

* conversion to hfst-optimized-lookup format was fundamentally broken in the
  original release; this should provide a fix and considerable speedup

* compose-intersect memory use is much improved

Noteworthy changes in 3.0
-------------------------

* all new API in object oriented form

* brand new foma backend for unweighted transducer handling

* the FST libraries used for automata handling are now dynamically linked to
  HFST

    * this requires end users to install needed backends themselves

* hfst-lexc and hfst-xfst based on foma's compiler

  * xfst is now also fully supported

  * the obsolete command line clients for HFST-based compilation of lexc and
    xfst scripts are hidden in hfst-lexc2fst and hfst-xfst2fst

* preliminary support for unknown character classes

* optimized-lookup backend supported in library

  * dropped special tools for optimised handling since optimised lookup is
    included in library: hfst-optimised-lookup -> hfst-lookup; 
    hfst-lookup-optimize -> hfst-fst2fst

* hfst-proc tool to emulate lt-proc, cg-proc

* Slightly improved string and text formats to support weights uniformly and
  parsing tabs and spaces as automata arcs: see `wiki page for text and string
  formats
  <https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstStringAndTextFormats>`

* better error handling using exceptions

* The argument handling order of binary tools changed: stdin is first if not
  specified

* The automata are now packed in HFST automaton container format (akin to AVI
  for videos); this supports arbitrary metadata and underlying FST formats

* the symbol handling is more implicit; no more symbol table arguments for tools
  or hfst-symbols command

* converted to new autotools: silent rules and color tests

* Hfst's internal automata format for fallback on some unsupported operations
  and for conversions

Noteworthy changes in 2.4.1
---------------------------

* Fixed inclusion path in pkg-config file

* Fixed extraneous epsilon path in hfst-lexc transducers

Noteworthy changes in release 2.4
---------------------------------


tools/src/hfst-fst2fst.cc
//! @file hfst-fst2fst.cc
//!
//! @brief Format conversion command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif


#include <iostream>
#include <fstream>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::HfstOutputStream;
using hfst::ImplementationType;


// tool-specific variables

ImplementationType output_type = hfst::UNSPECIFIED_TYPE;
bool hfst_format = true;
std::string options = "";

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
        "Convert transducers between binary formats\n"
        "\n", program_name);

    print_common_program_options(message_out);
    print_common_unary_program_options(message_out);
    fprintf(message_out, "Conversion options:\n"
    "  -f, --format=FMT                  Write result in FMT format\n"
    "  -b, --use-backend-format          Write result in implementation format, without any HFST wrappers\n"
    "  -S, --sfst                        Write output in (HFST's) SFST implementation\n"
    "  -F, --foma                        Write output in (HFST's) foma implementation\n"
    "  -t, --openfst-tropical            Write output in (HFST's) tropical weight (OpenFST) implementation\n"
    "  -l, --openfst-log                 Write output in (HFST's) log weight (OpenFST) implementation\n"
    "  -O, --optimized-lookup-unweighted Write output in the HFST optimized-lookup implementation\n"
    "  -w, --optimized-lookup-weighted   Write output in optimized-lookup (weighted) implementation\n"
    "  -Q  --quick                       When converting to optimized-lookup, don't try hard to compress\n");
    fprintf(message_out, "\n");
    print_common_unary_program_parameter_instructions(message_out);
        fprintf(message_out, 
            "FMT must be name of a format usable by libhfst, i.e. one of the following:\n"
        "{ openfst-tropical, openfst-log, sfst, foma,\n"
        "  optimized-lookup-weighted, optimized-lookup-unweighted }.\n");
    fprintf(message_out, "\n");
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}

int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          HFST_GETOPT_UNARY_LONG,
          // add tool-specific options here 
          {"use-backend-format", no_argument, 0, 'b'},
          {"format",       required_argument, 0, 'f'},
          {"sfst",               no_argument, 0, 'S'},
          {"foma",               no_argument, 0, 'F'},
          {"openfst-tropical",    no_argument, 0, 't'},
          {"openfst-log",         no_argument, 0, 'l'},
          {"optimized-lookup-unweighted",   no_argument, 0, 'O'},
          {"optimized-lookup-weighted",no_argument, 0, 'w'},
      {"quick",              no_argument, 0, 'Q'},
          {0,0,0,0}
        };
        int option_index = 0;
        // add tool-specific options here 
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "SFtlOwQf:b",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }


        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
          // add tool-specific cases here
                case 'f':
          output_type = hfst_parse_format_name(optarg);
          break;
        case 'b':
          hfst_format=false;
          break;
        case 'S':
          output_type = hfst::SFST_TYPE;
          break;
        case 'F':
          output_type = hfst::FOMA_TYPE;
          break;
        case 't':
          output_type = hfst::TROPICAL_OPENFST_TYPE;
          break;
        case 'l':
          output_type = hfst::LOG_OPENFST_TYPE;
          break;
        case 'O':
          output_type = hfst::HFST_OL_TYPE;
          break;
        case 'w':
          output_type = hfst::HFST_OLW_TYPE;
          break;
    case 'Q':
        options = "quick";
        break;
#include "inc/getopt-cases-error.h"
        }
    }
    
    if (output_type == hfst::UNSPECIFIED_TYPE)
    {
        error(EXIT_FAILURE, 0, 
              "You must specify an output type "
              "(one of -S, -f, -t, -l, -O, or -w)");
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    return EXIT_CONTINUE;
}
int
process_stream(HfstInputStream& instream, HfstOutputStream& outstream)
{
  //instream.open();
  //    outstream.open();
    
    size_t transducer_n = 0;
    while(instream.is_good())
    {
        transducer_n++;
        HfstTransducer orig(instream);
        char* inputname = hfst_get_name(orig, inputfilename);
        if (transducer_n == 1)
        {
          verbose_printf("Converting %s...\n", inputname); 
        }
        else
        {
          verbose_printf("Converting %s...%zu\n",
                         inputname, transducer_n);
        }
        try {
            orig.convert(output_type, options);
        } HFST_CATCH(HfstException)
        hfst_set_name(orig, orig, "convert");
        hfst_set_formula(orig, orig, "Id");
        outstream << orig;
    }
    instream.close();
    outstream.close();
    return EXIT_SUCCESS;
}


int main( int argc, char **argv ) {
    hfst_set_program_name(argv[0], "0.1", "HfstFst2Fst");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (inputfile != stdin)
    {
        fclose(inputfile);
    }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s\n", 
        inputfilename, outfilename);
    char* format_description = hfst_strformat(output_type);
    if (hfst_format) 
      {
        verbose_printf("Writing %s format transducers with HFST3 headers\n",
                       format_description);
      }
    else
      {
        verbose_printf("Writing %s format transducers without HFST specific"
                       " headers\n", format_description);
      }
    free(format_description);
    // here starts the buffer handling part
    HfstInputStream* instream = NULL;
    try {
      instream = (inputfile != stdin) ?
        new HfstInputStream(inputfilename) : new HfstInputStream();
    } catch(const HfstException e)  {
        error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
              inputfilename);
        return EXIT_FAILURE;
    }
    HfstOutputStream* outstream = (outfile != stdout) ?
      new HfstOutputStream(outfilename, output_type, hfst_format) :
      new HfstOutputStream(output_type, hfst_format);

    retval = process_stream(*instream, *outstream);
    delete instream;
    delete outstream;
    free(inputfilename);
    free(outfilename);
    return retval;
}

